src/bytecode.rs

```diff
@@ -0,0 +1,814 @@
+                self.instructions[call_iter_instr].set_end_target_call_iter_fn(end_target);
+                self.restore_continue_break(old_cont_break, end_target);
+                self.instructions.push(PopScope);
+            }
+
+            Stmt::Continue(_) => {
+                self.instructions.push(Branch{ target: self.continue_target });
+            },
+            Stmt::Break(_) => {
+                self.break_instrs.push(self.instructions.len());
+                self.instructions.push(Branch{ target: !0 });
+            }
+            Stmt::ReturnWithVal(expr, ReturnType::Return, _) => {
+                if let Some(expr) = expr {
+                    self.build_expr(expr)?
+                };
+                self.instructions.push(Return);
+            }
+            Stmt::ReturnWithVal(expr, ReturnType::Exception, _) => {
+                if let Some(expr) = expr {
+                    self.build_expr(expr)?;
+                }
+                else {
+                    self.instructions.push(StringConstant("".into()));
+                }
+
+                self.instructions.push(Panic);
+            }
+
+            Stmt::Let(name, Some(expr), _) => {
+                self.build_expr(expr)?;
+                self.instructions.push(CreateVariable{ name: name.clone() });
+            }
+
+            Stmt::Let(name, None, _) => {
+                self.instructions.push(PushUnit);
+                self.instructions.push(CreateVariable{ name: name.clone() });
+            }
+
+            Stmt::Const(name, expr, _) if expr.is_constant() => {
+                self.build_expr(expr)?;
+                self.instructions.push(CreateVariable{ name: name.clone() });
+            }
+
+            // TODO: Either this should be returning an error not panicking, or it
+            // should just be a debug_assert!. Same applies to the version of this
+            // in engine.rs::eval_stmnt
+            Stmt::Const(_, _, _) => panic!("constant expression not constant!"),
+        };
+
+        Ok(())
+    }
+
+    fn build_expr(&mut self, expr: &Expr) -> Result<(), BuildError> {
+        match *expr {
+            Expr::True(_) => self.instructions.push(TrueConstant),
+            Expr::False(_) => self.instructions.push(FalseConstant),
+            Expr::Unit(_) => self.instructions.push(PushUnit),
+            Expr::IntegerConstant(i, _ ) =>
+                self.instructions.push(IntegerConstant(i)),
+            Expr::FloatConstant(f, _) =>
+                self.instructions.push(FloatConstant(f)),
+            Expr::StringConstant(ref s, _) =>
+                self.instructions.push(StringConstant(s.clone())),
+            Expr::CharConstant(c, _) =>
+                self.instructions.push(CharConstant(c)),
+
+            // TODO: Removed if !state.always_search
+            Expr::Variable(_, Some(index), _) =>
+                self.instructions.push(GetVariable{ index }),
+            // TODO: Is this ever used without eval?
```

Comment:
You're right.  `always_search` is only ever set to `true` when there is an `eval` call and the number of variables in the scope changes.

Reply:
Sort of discussed below - disallowing/backing out to the AST version of the code seems like the simplest solution (which means outright disallowing in functions that want the ability to yield/call functions with the ability to yield - but as originally mentioned I'm not sure that's a particularly mainstream feature)